tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/CloneRepos.cs

```diff
@@ -0,0 +1,155 @@
+            p.WaitForExit();
+            if (p.ExitCode == 0)
+            {
+                Log.LogMessage("git command", null, null, null, 0, 0, 0, 0, MessageImportance.Low, await p.StandardOutput.ReadToEndAsync());
+                return true;
+            }
+            else
+            {
+                Log.LogError("git command", p.ExitCode, null, null, 0, 0, 0, 0, $"command {psi.FileName} {psi.Arguments} in directory {psi.WorkingDirectory} failed.  Command output follows:");
+                Log.LogError(await p.StandardOutput.ReadToEndAsync());
+                Log.LogError(await p.StandardError.ReadToEndAsync());
+                return false;
+            }
+        }
+
+        private async Task<bool> CheckoutRepo(string path, string hash)
+        {
+            return await RunExternalGitCommand($"checkout {hash}") &&
+                   await RunExternalGitCommand($"submodule update --init --recursive");
+        }
+
+        private async Task<bool> CloneRepo(string url, string path)
+        {
+            return await RunExternalGitCommand($"clone --recursive {url} {path}");
+        }
+
+        private async void EnsureGit()
+        {
+            try
+            {
+                var gitExists = await RunExternalGitCommand("--version");
+            }
+            catch (Exception e)
+            {
+                Log.LogError("git does not appear to be installed or working.  See error message for details.");
+                Log.LogErrorFromException(e, true, true, null);
+                throw;
+            }
+        }
+
+        private static IEnumerable<RepoSpec> ReadSpecFile(string filePath)
+        {
+            var doc = XDocument.Load(filePath);
+            foreach (var dep in doc.Element("Dependencies").Element("ToolsetDependencies").Elements("Dependency").Concat(doc.Element("Dependencies").Element("ProductDependencies").Elements("Dependency")))
+            {
+                if (!string.IsNullOrWhiteSpace(dep.Element("SkipClone")?.Value) && dep.Element("SkipClone").Value.ToLowerInvariant() == "true")
+                {
+                    continue;
+                }
+                if (!string.IsNullOrWhiteSpace(dep.Element("RepoName")?.Value))
+                {
+                    yield return new RepoSpec(dep.Element("Uri").Value, dep.Element("RepoName").Value, dep.Element("Sha").Value);
+                }
+                else
+                {
+                    yield return new RepoSpec(dep.Element("Uri").Value, dep.Element("Sha").Value);
+                }
+            }
+        }
+
+        private class RepoSpec
+        {
+            internal string Url { get; set; }
+            internal string Name { get; set; }
+            internal string Hash { get; set; }
+
+            public RepoSpec(string url, string hash) : this(url, url.Substring(url.LastIndexOf('/') + 1), hash)
+            {
+            }
+
+            public RepoSpec(string url, string name, string hash)
```

Comment:
Same-positioned args should have the same meaning especially if the type is the same.